src/unix_term.rs

```diff
@@ -66,41 +65,76 @@ pub(crate) fn terminal_size(out: &Term) -> Option<(u16, u16)> {
     }
 }
 
-pub(crate) fn read_secure() -> io::Result<String> {
-    let mut f_tty;
-    let fd = unsafe {
-        if libc::isatty(libc::STDIN_FILENO) == 1 {
-            f_tty = None;
-            libc::STDIN_FILENO
+enum Input<T> {
+    Stdin(io::Stdin),
+    File(T),
+}
+
+impl Input<fs::File> {
+    fn new() -> io::Result<Self> {
+        let stdin = io::stdin();
+        if is_a_terminal(&stdin) {
+            Ok(Input::Stdin(stdin))
         } else {
             let f = fs::OpenOptions::new()
                 .read(true)
                 .write(true)
                 .open("/dev/tty")?;
-            let fd = f.as_raw_fd();
-            f_tty = Some(BufReader::new(f));
-            fd
+            Ok(Input::File(f))
         }
-    };
+    }
+}
+
+impl Input<BufReader<fs::File>> {
+    fn new() -> io::Result<Self> {
+        Ok(match Input::<fs::File>::new()? {
+            Input::Stdin(s) => Self::Stdin(s),
+            Input::File(f) => Self::File(BufReader::new(f)),
+        })
+    }
+}
+
+impl<T: BufRead> Input<T> {
+    fn read_line(&mut self, buf: &mut String) -> io::Result<usize> {
+        match self {
+            Self::Stdin(s) => s.read_line(buf),
+            Self::File(f) => f.read_line(buf),
+        }
+    }
+}
+
+impl AsRawFd for Input<fs::File> {
+    fn as_raw_fd(&self) -> RawFd {
+        match self {
+            Self::Stdin(s) => s.as_raw_fd(),
+            Self::File(f) => f.as_raw_fd(),
+        }
+    }
+}
+
+impl AsRawFd for Input<BufReader<fs::File>> {
+    fn as_raw_fd(&self) -> RawFd {
+        match self {
+            Self::Stdin(s) => s.as_raw_fd(),
+            Self::File(f) => f.get_ref().as_raw_fd(),
+        }
+    }
+}
+
+pub(crate) fn read_secure() -> io::Result<String> {
+    let mut input = Input::<BufReader<fs::File>>::new()?;
 
     let mut termios = mem::MaybeUninit::uninit();
-    c_result(|| unsafe { libc::tcgetattr(fd, termios.as_mut_ptr()) })?;
+    c_result(|| unsafe { libc::tcgetattr(input.as_raw_fd(), termios.as_mut_ptr()) })?;
     let mut termios = unsafe { termios.assume_init() };
     let original = termios;
     termios.c_lflag &= !libc::ECHO;
-    c_result(|| unsafe { libc::tcsetattr(fd, libc::TCSAFLUSH, &termios) })?;
+    c_result(|| unsafe { libc::tcsetattr(input.as_raw_fd(), libc::TCSAFLUSH, &termios) })?;
     let mut rv = String::new();
 
-    let read_rv = if let Some(f) = &mut f_tty {
-        f.read_line(&mut rv)
-    } else {
-        io::stdin().read_line(&mut rv)
-    };
-
-    c_result(|| unsafe { libc::tcsetattr(fd, libc::TCSAFLUSH, &original) })?;
+    let read_rv = input.read_line(&mut rv);
 
-    // Ensure the fd is only closed after everything has been restored.
-    drop(f_tty);
```

Comment:
I think this is a rebase error?

Reply:
Nope. This code was here to ensure that `f_tty` outlives the last use of `fd`, since both represent the same underlying resource whose lifetime is tied to `f_tty`. This is also the reason I won't write `let fd = ...`; calling `as_raw_fd()` is trivial but ensures that the proper lifetimes are preserved.